### DISCUSSION TITLE
Factory Pattern For Smart Contract Using JS

### DISCUSSION THREAD

**Discussion by hiraqdev** [near/near-sdk-js#185] [Q&A] [answered]
I'm newcomer to NEAR , before i'm develop a smart contract at Ethereum using Solidity, and i've been familiar with _factory pattern_ from there, just for simple explanation about _factory pattern_ itself is we can "clone" smart contract with different contract's address, we can _deploy_ multiple contract's instance from some contract.  

I'm using this pattern when found a usecase when a user need their own contract.  One user one contract, and it's specific only for a single user, for example maintain user's inventory.  Actually the inventory contract itself is always same between users, but rather than a single inventory contract that maintain _mapping_ for each users, we just deploy an inventory contract for each user, so each user has their own inventory.

Ref: 
- https://betterprogramming.pub/learn-solidity-the-factory-pattern-75d11c3e7d29
- https://blog.logrocket.com/cloning-solidity-smart-contracts-factory-pattern/

The question is, is it possible to implement this pattern at NEAR's contract ? Of course using `near-sdk-js` , because i've just learned the NEAR's contract behaviors, and it looks like we need to deploy the contract manually via command line, or maybe i'm just wrong?

**Comment by ailisp** [comment, marked as answer]
Yes. You can refer to test_promise_batch_deploy_call method of https://github.com/near/near-sdk-js/blob/develop/tests/src/promise_batch_api.js function. It's a contract that deploying the given contract bytes. 
The example looks ugly - have to provide the other contract's binary content (wasm) in a JS string and that can be extremely long. We're going to add a utility to allow `include(filename)` in the build process to make it a lot cleaner

**Reply by hiraqdev** [reply]
is it same like `include_bytes` from Rust?

**Reply by ailisp** [reply]
Yes!
